Schema based annotation

kkunal1408 opened the pull request.

kkunal1408 commented:
@parijatm Not sure why the wheel is crashing.

parijatm commented:
Looks like a boost download issue.

kkunal1408 commented:
@soneryaldiz Please find my intermediate updates here.

parijatm commented:
Master has changed significantly since this PR was opened @kkunal1408. I would suggest rebasing / merging before you continue making changes. In addition, it may be a good idea to come up with a mapping from the new schema to the hacked schema I had to create to get constraint checking working. This mapping step can then be moved later and later in the flow to incrementally go from ALIGN v1 to v2. Please set up a meeting if you would like to discuss this further.

@soneryaldiz this PR will take some time to fully merge so please go ahead and make whatever changes you need to get 'generate' working.

kkunal1408 commented:
@parijatm can you start looking at this PR. This is almost ready with.  (Need to fix the power pin deletion requirement before PnR)

parijatm commented:
> @parijatm can you start looking at this PR. This is almost ready with. (Need to fix the power pin deletion requirement before PnR)

Hi @kkunal1408. I looked through the flow in align.compiler.compiler and **love** what I see so far.  

There are some changes I would like to see in the final flow but would like to deal with them in separate PRs. Just making a list for future reference as some of them fall outside of annotation as well:
1. Replacing generators.json with a python function that looks for generator classes
2. Making model.txt, basic_library.sp etc. pdk specific
3. Ability to use the PDK library models in ckt without re-parsing model.txt
4. Revisit user_template.sp to see if it can be replaced with constraints / generators
5. Re-use schema directly in PnR instead of writing a separate VerilogJson data structure
6. Create an "Engine" class & write each transformation as a subclass of the same.

I will try and work on (5) and (6) once you get this merged in. Please get CI passing and run the entire regression suite as well to make sure we didn't miss anything. I will try and do a more detailed review as well over the next couple of days.

kkunal1408 commented:
@Lastdayends Can you please help in fixing the DRCs in the adder circuit?

kkunal1408 commented:
Ready for review. @ soner can you please help me looking into at the last test. It is passing on my local system.

Lastdayends commented:
@kkunal1408 @854768750 

Is the DRC gone?

kkunal1408 commented:
> @kkunal1408 @854768750
> 
> Is the DRC gone?

No it still has DRCs. Just that I have turned it off in CI.

Lastdayends commented:
@kkunal1408 

I see.

@854768750 

Could you help to have a look?

854768750 commented:
ok, I am looking into `tests/pnr/test_entrypoint2.py::test_matrix`

854768750 commented:
@kkunal1408 I got the same result as in CI. The problem is I cannot get a copy of the input files in the tests/ folder. Have you managed to get the input files? It will be easier to debug this test case.

854768750 commented:
I found this test case is good in the master branch. Should we merge the master branch into this branch?

stevenmburns commented:
I just did the merge locally and cleaned up the collisions. I'll push it now.

854768750 commented:
> @kkunal1408 I got the same result as in CI. The problem is I cannot get a copy of the input files in the tests/ folder. Have you managed to get the input files? It will be easier to debug this test case.

Please ignore this comment. The input files are copied to `ALIGN_WORK_DIR / f'{nm}_entrypoint2'`

kkunal1408 commented:
> @kkunal1408 I got the same result as in CI. The problem is I cannot get a copy of the input files in the tests/ folder. Have you managed to get the input files? It will be easier to debug this test case.

I wanted help in adder testcase from  examples directory. Matrix testcase I can look into.

stevenmburns commented:
The matrix case completely bypasses Kunal's front end so I don't think we need to worry about it.

These are failing now:
```
---------------------------------------------------------------- generated xml file: /home/smburns/DARPA/ALIGN-public/junit.xml ----------------------------------------------------------------
=================================================================================== short test summary info ====================================================================================
FAILED tests/compiler/test_user_constraint.py::test_group_block_hsc[high_speed_comparator_orderblock] - FileExistsError: [Errno 17] File exists: '/home/smburns/DARPA/ALIGN-public/tests/comp...
FAILED tests/compiler/test_user_constraint.py::test_group_block_hsc[high_speed_comparator_symmblock] - FileExistsError: [Errno 17] File exists: '/home/smburns/DARPA/ALIGN-public/tests/compi...
FAILED tests/compiler/test_user_constraint.py::test_group_block_hsc[high_speed_comparator_portlocation] - FileExistsError: [Errno 17] File exists: '/home/smburns/DARPA/ALIGN-public/tests/co...
FAILED tests/compiler/test_user_constraint.py::test_group_block_hsc[high_speed_comparator_multiconnection] - FileExistsError: [Errno 17] File exists: '/home/smburns/DARPA/ALIGN-public/tests...
FAILED tests/compiler/test_user_constraint.py::test_group_block_hsc[high_speed_comparator_align] - FileExistsError: [Errno 17] File exists: '/home/smburns/DARPA/ALIGN-public/tests/compiler/...
FAILED tests/compiler/test_user_constraint.py::test_group_block_hsc[high_speed_comparator_symmnet] - FileExistsError: [Errno 17] File exists: '/home/smburns/DARPA/ALIGN-public/tests/compile...
FAILED tests/pnr/test_entrypoint2.py::test_row - assert None is not None
FAILED tests/pnr/test_entrypoint2.py::test_matrix - assert None is not None
```

I'll look into the entrypoint2 stuff.

854768750 commented:
@kkunal1408 For "adder" test case, I got two opens. Did you get the same result?

kkunal1408 commented:
> @kkunal1408 For "adder" test case, I got two opens. Did you get the same result?

Yes,

854768750 commented:
![picture](https://user-images.githubusercontent.com/28666144/128424062-f2a6cab4-9c56-403c-91af-566ba2a8fc34.PNG)
I found a problem. The open happens when routing cap pins. And the cap pins are out of the block boundary.
![picture1](https://user-images.githubusercontent.com/28666144/128424594-709401fe-7cf1-4174-a3a9-99cda88f28e2.PNG)
And these two pins should be connected. But their distance is only one M2 pitch

kkunal1408 commented:
Thanks Yishuang. @arvuce22 can you please help in fixing the cap layout?

854768750 commented:
> Thanks Yishuang. @arvuce22 can you please help in fixing the cap layout?

But I also tried to separate the pins to 3 M2 pitches, there is still open. Let me see what is the reason.

stevenmburns commented:
It looks like it might have trouble routing between the pins even if they are spaced farther apart. The top blue pin (on the resistor) has to connect to the blue pin below connected to the capacitor. There are likely obstructions there because of the m3 in the capacitor. It would have to do wrong way routing to make the connection and avoid the m3 obstructions. Do you have a visualization of what the global router does in this case?

![Screenshot from 2021-08-05 16-06-36](https://user-images.githubusercontent.com/1512574/128432599-986d5fc2-f025-4231-9256-32748961232d.png)

854768750 commented:
I see. Those M3 are obstruction in lef.

stevenmburns reviewed (changes requested):
A few things to start with:
1) We have print statements (not even log statement) showing up. "__root__=...."
2) For some reason the name of all the blocks are now capitalized. I don't think we should be changing names. Why do we need to do this?

stevenmburns commented on align/main.py:
```diff
@@ -287,7 +320,7 @@ def schematic2layout(netlist_dir, pdk_dir, netlist_file=None, subckt=None, worki
         netlist = extract_netlist_files(netlist_dir,netlist_file)
         if subckt is None:
-            subckt = netlist.stem
+            subckt = netlist.stem.upper()
```
Why do we need to do this? (And everywhere else.) If we have to do it, we should do it in one place.

kkunal1408 replied:
The new netlist parser from schema is case insensitive (similar to spice syntax assumption) and the idea is to run the whole flow in caps instead of worrying about case.

stevenmburns commented on align/pdk/finfet/transistor_array.py:
```diff
@@ -31,33 +31,34 @@ def mos_array_temporary_wrapper(self, x_cells, y_cells, pattern, vt_type, ports,
+            assert False, f'Either nf>1 or stack>1 parameter should be defined {parameters}'
+
+        if 'W' in parameters:
+            nfin = int(float(parameters['W']) * 1e10) // self.pdk['Fin']['Pitch']
```
What should happen here if 'W' is not divisible by the fin pitch? Currently is silently truncates. It should probably be an error or maybe round to nearest. Truncate seems wrong.

kkunal1408 replied:
This is a hack added by Soner for intel 22fl pdk.

soneryaldiz replied:
yes, this was intentional as the parsed floating point numbers had long tail.

stevenmburns replied:
You'll probably want to replace this with:
```
w = round(float(parameters['W']) * 1e10)
nfin,r = divmod(w, self.pdk['Fin']['Pitch'])
if r != 0:
   logger.warning( f"nfin not integral: {w} // {self.pdk['Fin']['Pitch']} = {nfin} with remainder {r}")
```
Otherwise 2.9999 might round to 2 instead of 3 and also w in anstroms might now be a multiple of the fin pitch [if we are worried that W is not an integral number of angstroms, we need to do more here as well.]
We could make this check be in a schema validator.

stevenmburns commented:
@kkunal1408 Please let me know which parts of this you want to have reviewed in detail. Are there parts that you want advice on?

kkunal1408 commented:
> A few things to start with:
> 
> 1. We have print statements (not even log statement) showing up. "**root**=...."
Fixed
> 2. For some reason the name of all the blocks are now capitalized. I don't think we should be changing names. Why do we need to do this?
Enforced by schema and the parser, for which most of the code was written by Parijat. Do not want to touch it in this pr. I have removed any dependencies outside complier on this. We can change it back to lowercase later.

stevenmburns commented:
@854768750 @Lastdayends It seems the "-pdn" mode is failing (seg fault) for the telescopic_ota example now.
There is a test for this but if you want to run in in isolation:
```
schematic2layout.py ../../examples/telescopic_ota -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS -pdn
```
The lowest level metal is MAXINT which is causing a segfault. it seems that we are no longer finding the power pins which is leading to this problem.
Here is the crash point:
```
Thread 1 "python3.8" received signal SIGSEGV, Segmentation fault.
0x00007fffd4494336 in MNASimulation::Map_new (this=0x31d2950, currentstore=std::vector of length 20, capacity 32 = {...}, point_set=std::set with 0 elements, 
    Power_Grid_devices=std::vector of length 0, capacity 0, metal_layer=2147483647) at ../../../PlaceRouteHierFlow/MNA/MNASimulation.cpp:1286
1286	    double unit_r = this->Drc_info.Metal_info[metal_layer].unit_R;
```
but the problem occurs but the stack, here probably:
```
941	   int highest_metal = INT_MIN;
942	   int lowest_metal = INT_MAX;
943	   double VDD = 0.8;
944	
945	   ExtractPowerGridPoint(vdd, point_set, highest_metal, lowest_metal, VDD);
946	   ExtractPowerGridPoint(gnd, point_set, highest_metal, lowest_metal, 0.0);
```
It seems that something is not found due to a change in the compiler code.

stevenmburns commented:
I guess we can see the problem even earlier because no power routing is generated:
```
schematic2layout.py ../../examples/telescopic_ota -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS -pdn
...
align.compiler.user_const INFO : No user constraints found for block TELESCOPIC_OTA in path /home/smburns/DARPA/ALIGN-public/examples/telescopic_ota
align.compiler.compiler INFO : Power and ground nets not found. Power grid will not be constructed.
align.compiler.compiler INFO : Completed topology identification.
...
```

@kkunal1408 Maybe this is something you should fix first before the PnR looks at the problem. (Although the PnR shouldn't crash in the case no matter what.)

stevenmburns commented:
@kkunal1408 I thought I could fix this by changing the power net definitions but nothing worked. Here is what I did last:
```
==============telescopic_ota.sp============
.subckt telescopic_ota vbiasn vbiasp1 vbiasp2 vinn vinp voutn voutp vss vdd
.param no_of_fin = 10
m1 id id vss vss nmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m2 net10 id vss vss nmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m5 voutn vbiasn net8 vss nmos_rvt w=270e-9 l=20e-9 nfin=10
m6 voutp vbiasn net014 vss nmos_rvt w=270e-9 l=20e-9 nfin=10
m8 voutp vbiasp1 net012 vdd pmos_rvt w=270e-9 l=20e-9 nfin=15
m7 voutn vbiasp1 net06 vdd pmos_rvt w=270e-9 l=20e-9 nfin=15
m10 net012 vbiasp2 vdd vdd pmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m9 net06 vbiasp2 vdd vdd pmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m4 net014 vinn net10 vss nmos_rvt w=270e-9 l=20e-9 nfin=75
m3 net8 vinp net10 vss nmos_rvt w=270e-9 l=20e-9 nfin=75
.ends ota
** End of subcircuit definition.
=====================telescopic_ota.setup==========================
.subckt telescopic_ota vbiasn vbiasp1 vbiasp2 vinn vinp voutn voutp vss vdd
.param no_of_fin = 10
m1 id id vss vss nmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m2 net10 id vss vss nmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m5 voutn vbiasn net8 vss nmos_rvt w=270e-9 l=20e-9 nfin=10
m6 voutp vbiasn net014 vss nmos_rvt w=270e-9 l=20e-9 nfin=10
m8 voutp vbiasp1 net012 vdd pmos_rvt w=270e-9 l=20e-9 nfin=15
m7 voutn vbiasp1 net06 vdd pmos_rvt w=270e-9 l=20e-9 nfin=15
m10 net012 vbiasp2 vdd vdd pmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m9 net06 vbiasp2 vdd vdd pmos_rvt w=270e-9 l=20e-9 nfin=15 nf=2
m4 net014 vinn net10 vss nmos_rvt w=270e-9 l=20e-9 nfin=75
m3 net8 vinp net10 vss nmos_rvt w=270e-9 l=20e-9 nfin=75
.ends ota
** End of subcircuit definition.
```
I still get that no power grid is being generated. I suspect it is never generated now. Maybe that is the problem.

stevenmburns commented:
@kkunal1408 Traditional things that worked are failing now:
```
 schematic2layout.py ../../examples/switched_capacitor_filter/ -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS
align.main INFO : Running flow steps ['1_topology', '2_primitives', '3_pnr:prep', '3_pnr:place', '3_pnr:route', '3_pnr:check']
align.main INFO : READ file: /home/smburns/DARPA/ALIGN-public/examples/switched_capacitor_filter/switched_capacitor_filter.sp subckt=SWITCHED_CAPACITOR_FILTER, flat=0
align.compiler.compiler INFO : Starting topology identification...
align.compiler.compiler INFO : Reading setup file /home/smburns/DARPA/ALIGN-public/examples/switched_capacitor_filter/switched_capacitor_filter.setup
align.schema.constraint INFO : reading data {'constraint': 'GroupCaps', 'instances': ['c1', 'c3'], 'name': 'c1_c3', 'unit_cap': 'Cap_12f', 'num_units': [3, 5], 'dummy': False}
align.schema.constraint INFO : reading data {'constraint': 'GroupCaps', 'instances': ['c1', 'c3'], 'name': 'c1_c3', 'unit_cap': 'Cap_12f', 'num_units': [3, 5], 'dummy': False}
align.cmdline ERROR : Fatal Error. Cannot proceed
Traceback (most recent call last):
  File "/home/smburns/DARPA/ALIGN-public/align/cmdline.py", line 142, in parse_args
    return schematic2layout(**vars(arguments))
  File "/home/smburns/DARPA/ALIGN-public/align/main.py", line 329, in schematic2layout
    primitives = generate_hierarchy(netlist, subckt, topology_dir, flatten, pdk_dir, uniform_height)
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 24, in generate_hierarchy
    ckt_data = compiler_input(netlist_path, subckt, pdk_dir, config_path, flatten_heirarchy)
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 116, in compiler_input
    annotate.annotate()
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/match_graph.py", line 75, in annotate
    self._group_cap_const(circuit_name)
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/match_graph.py", line 208, in _group_cap_const
    and hasattr(const, 'instances') and len(const_inst) > 1:
UnboundLocalError: local variable 'const_inst' referenced before assignment
```

kkunal1408 commented:
> schematic2layout.py ../../examples/telescopic_ota -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS -pdn

This works for me. Can you pull the latest code?

kkunal1408 commented:
> @kkunal1408 Traditional things that worked are failing now:
> 
> ```
>  schematic2layout.py ../../examples/switched_capacitor_filter/ -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS
> align.main INFO : Running flow steps ['1_topology', '2_primitives', '3_pnr:prep', '3_pnr:place', '3_pnr:route', '3_pnr:check']
> align.main INFO : READ file: /home/smburns/DARPA/ALIGN-public/examples/switched_capacitor_filter/switched_capacitor_filter.sp subckt=SWITCHED_CAPACITOR_FILTER, flat=0
> align.compiler.compiler INFO : Starting topology identification...
> align.compiler.compiler INFO : Reading setup file /home/smburns/DARPA/ALIGN-public/examples/switched_capacitor_filter/switched_capacitor_filter.setup
> align.schema.constraint INFO : reading data {'constraint': 'GroupCaps', 'instances': ['c1', 'c3'], 'name': 'c1_c3', 'unit_cap': 'Cap_12f', 'num_units': [3, 5], 'dummy': False}
> align.schema.constraint INFO : reading data {'constraint': 'GroupCaps', 'instances': ['c1', 'c3'], 'name': 'c1_c3', 'unit_cap': 'Cap_12f', 'num_units': [3, 5], 'dummy': False}
> align.cmdline ERROR : Fatal Error. Cannot proceed
> Traceback (most recent call last):
>   File "/home/smburns/DARPA/ALIGN-public/align/cmdline.py", line 142, in parse_args
>     return schematic2layout(**vars(arguments))
>   File "/home/smburns/DARPA/ALIGN-public/align/main.py", line 329, in schematic2layout
>     primitives = generate_hierarchy(netlist, subckt, topology_dir, flatten, pdk_dir, uniform_height)
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 24, in generate_hierarchy
>     ckt_data = compiler_input(netlist_path, subckt, pdk_dir, config_path, flatten_heirarchy)
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 116, in compiler_input
>     annotate.annotate()
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/match_graph.py", line 75, in annotate
>     self._group_cap_const(circuit_name)
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/match_graph.py", line 208, in _group_cap_const
>     and hasattr(const, 'instances') and len(const_inst) > 1:
> UnboundLocalError: local variable 'const_inst' referenced before assignment
> ```

Working on this. Based on the latest discussions from Meghna, the Groupcap constraint was supposed to be removed from switched capacitor filter.

kkunal1408 commented:
> @kkunal1408 Please let me know which parts of this you want to have reviewed in detail. Are there parts that you want advice on?

Mainly the interaction between schema and compiler. And any changes in schema should be looked at closely. This forms the baseline. Some Features were supposed to be modified later.

- Lots of TODOs are listed in the compiler code for the improvements.

stevenmburns commented:
> > schematic2layout.py ../../examples/telescopic_ota -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS -pdn
> 
> This works for me. Can you pull the latest code?

This still fails for me after a git pull.

kkunal1408 commented:
> > > schematic2layout.py ../../examples/telescopic_ota -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS -pdn
> > 
> > 
> > This works for me. Can you pull the latest code?
> 
> This still fails for me after a git pull.

Are you seeing this line after you run it: "align.compiler.compiler INFO : Reading setup file /mnt/d/research_work/ALIGN/mywork/ALIGN-public/examples/telescopic_ota/telescopic_ota.setup"
In the new compiler, the requirement is that the subcircuit definition should have power pins.
Are you able to generate pdn for any other example such as buffer?
Or can we recreate the issue on CI?

stevenmburns reviewed (changes requested):
Some things still need to be fixed.

stevenmburns commented:
> > > > schematic2layout.py ../../examples/telescopic_ota -p ../../pdks/FinFET14nm_Mock_PDK/ --skipGDS -pdn
> > > 
> > > 
> > > This works for me. Can you pull the latest code?
> > 
> > 
> > This still fails for me after a git pull.
> 
> In the new compiler, the requirement is that the subcircuit definition should have power pins.
> 
> * Are you seeing this line after you run it: "align.compiler.compiler INFO : Reading setup file /mnt/d/research_work/ALIGN/mywork/ALIGN-public/examples/telescopic_ota/telescopic_ota.setup" ?
> * Are you able to generate pdn for any other example such as buffer?
> * Or can we recreate the issue on CI?

I found this issue diagnosing your second CI failure. It was this one.

![image](https://user-images.githubusercontent.com/1512574/128568575-51cfb105-3db9-4dcf-a7c4-8d366b4d1840.png)

As far as I can tell now, no runs are generating power grids (this test is failing because it needs a power grid to do the power grid simulation.)

stevenmburns commented:
@kkunal1408 Same problem for the buffer as well. No power pins being generated.

kkunal1408 commented:
@854768750 Can you help me debug this crash in high_speed_comparator design?
tests/test_cmdline.py::test_cmdline[high_speed_comparator-1-False] 
This design passes on my and ramprasath's systems.

854768750 commented:
> @854768750 Can you help me debug this crash in high_speed_comparator design?
> tests/test_cmdline.py::test_cmdline[high_speed_comparator-1-False]
> This design passes on my and ramprasath's systems.

Let me have a look.

kkunal1408 commented:
@stevenmburns Can we set up CI so that it stores logs for failing runs? The runs which pass on personal systems but fail on CI are difficult to debug.

kkunal1408 commented:
> I see. Those M3 are obstruction in lef.

Did we find a solution for this?

854768750 commented:
> > I see. Those M3 are obstruction in lef.
> 
> Did we find a solution for this?

Is it possible to move primitive pins to the top metal layer? Such as moving M2 pins to M4 or moving M3 obstructions to M1?

stevenmburns commented:
@srini229 @854768750 I'm having trouble running any `schematric2layout.py` run through when this branch is complied using a build type of `Release` or `RelWithDebInfo`.
I'm getting a core dump here:
```
#0  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fffffffb398, __in_chrg=<optimized out>) at /usr/include/c++/7/bits/shared_ptr_base.h:684
#1  std::__shared_ptr<spdlog::logger, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fffffffb390, __in_chrg=<optimized out>) at /usr/include/c++/7/bits/shared_ptr_base.h:1123
#2  std::shared_ptr<spdlog::logger>::~shared_ptr (this=0x7fffffffb390, __in_chrg=<optimized out>) at /usr/include/c++/7/bits/shared_ptr.h:93
#3  ILP_solver::GenerateValidSolution (this=this@entry=0x7fffffffb930, mydesign=..., curr_sp=..., drcInfo=...) at ../../../PlaceRouteHierFlow/placer/ILP_solver.cpp:67
#4  0x00007fffd4543d01 in Placer::PlacementCoreAspectRatio_ILP (this=this@entry=0x7fffffffc0c0, designData=..., curr_sp=..., curr_sol=..., mode=mode@entry=0, nodeSize=<optimized out>, 
    effort=<optimized out>, drcInfo=...) at ../../../PlaceRouteHierFlow/placer/Placer.cpp:635
#5  0x00007fffd4546d55 in Placer::PlacementRegularAspectRatio_ILP (this=this@entry=0x7fffffffc0c0, nodeVec=std::vector of length 1, capacity 1 = {...}, opath="./Results/", 
    effort=effort@entry=0, drcInfo=...) at ../../../PlaceRouteHierFlow/placer/Placer.cpp:739
#6  0x00007fffd45472cb in Placer::Placer (this=0x7fffffffc0c0, nodeVec=std::vector of length 1, capacity 1 = {...}, opath=..., effort=0, drcInfo=..., hyper_in=...)
    at ../../../PlaceRouteHierFlow/placer/Placer.cpp:18
```
Valgrind tells me there is a problem with an address:
```
align.pnr.toplevel INFO : Starting bottom-up placement on STAGE2_INV_PG0 2
PnR.placer.SeqPair.SetEnumerate INFO : Enumerated search
==13815== Warning: client switching stacks?  SP change: 0x1ffeffd060 --> 0x40b6f80000000000
==13815==          to suppress, use: --max-stackframe=4663187005650251680 or greater
==13815== Warning: client switching stacks?  SP change: 0x40b6f80000000000 --> 0x1ffeffd308
==13815==          to suppress, use: --max-stackframe=4663187005650251000 or greater
==13815== Warning: client switching stacks?  SP change: 0x1ffeffd060 --> 0x40c6f80000000000
==13815==          to suppress, use: --max-stackframe=4667690605277622176 or greater
==13815==          further instances of this message will not be shown.
==13815== Invalid write of size 8
==13815==    at 0x2D3DA90D: ILP_solver::GenerateValidSolution(design&, SeqPair const&, PnRDB::Drc_info const&) (ILP_solver.cpp:494)
==13815==  Address 0x40c6f7ffffffffe8 is not stack'd, malloc'd or (recently) free'd
==13815== 
==13815== 
```
Not working in Release mode is something we have to fix. It might also explain why we are getting different behaviors on different developers individual systems (even in Debug mode.)

854768750 commented:
> test_cmdline

This design is also correct on my computer using pytest.

stevenmburns commented:
@kkunal1408 I merged in the array out of bounds error (on the stack) that was causing problems from before.
There are still a lot of pytest failures for me. Do you want me to work on them one by one, or are you doing that?

kkunal1408 commented:
Thanks steve. I will give a shot at fixing these as I can see these errors
on my system.
Kunal


On Mon, Aug 16, 2021 at 2:35 PM Steve Burns ***@***.***>
wrote:

> @kkunal1408 <https://github.com/kkunal1408> I merged in the array out of
> bounds error (on the stack) that was causing problems from before.
> There are still a lot of pytest failures for me. Do you want me to work on
> them one by one, or are you doing that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ALIGN-analoglayout/ALIGN-public/pull/615#issuecomment-899833705>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJUB53JNNF3U5QITUROLAT3T5GACNANCNFSM42W64INQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

kkunal1408 commented:
Hi Steve/Parijat
One of the python checks is not working on CI, which I need help with.
How flow should work: 

1. Find matches of primitive (e.g. DP) in a circuit
2. Each primitive has a constraint associated with it. Check if the primitive constraints will be valid for the matching nodes in the circuit.
3. If the constraints are valid for the matching nodes, create a primitive hierarchy using the nodes of the circuit.
4. If the constraint is not valid for matching nodes, discard the match.

For the FAILED tests/compiler/test_sizing.py::test_sizing3[intel_circuit3]
On CircleCI the second step of the check is not happening, while it works fine on my system.
This is the print statement on CI:
INFO align.schema.graph:graph.py:118 checking symmblock constraint of the subcircuit DP_NMOS_B __root__=[List[str](__root__=['M0', 'M1'])] at INTEL_CIRCUIT3

After this statement this line should be executed but its not happening: 
align/schema/constraint.py(line 535):logger.info(f"Checking symmetric block constr for {pair} {self.parent.parent}")

DP should not match here as the device sizes are different. 
align/schema/constraint.py (line 539) :assert self.parent.parent.get_element(pair[0]).parameters == \
                    self.parent.parent.get_element(pair[1]).parameters

stevenmburns commented:
It seems that the z3-solver isn't  installed on that particular setup and your code depends on it being installed to produce the correct answer.

This is what I get on my local machine when I uninstall z3-solver:
```
(.venv) smburns@smburns-XPS-13-9370:~/DARPA/ALIGN-public$ pytest -s -k test_sizing3 -vv -- tests/compiler/test_sizing.py
align.schema.checker WARNING : Could not import z3. Z3Checker disabled.
======================================================================================= test session starts =======================================================================================
platform linux -- Python 3.8.0, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/smburns/DARPA/ALIGN-public/.venv/bin/python3.8
cachedir: .pytest_cache
rootdir: /home/smburns/DARPA/ALIGN-public, configfile: pytest.ini
plugins: cov-2.12.1, forked-1.3.0, timeout-1.4.2, cpp-2.0.0, dash-1.20.0, xdist-2.2.1
collected 5 items / 4 deselected / 1 selected                                                                                                                                                     

tests/compiler/test_sizing.py::test_sizing3[intel_circuit3] align.compiler.preprocess WARNING : No power and gnd in this circuit INTEL_CIRCUIT3, please check setup file
intel_circuit3
name='INTEL_CIRCUIT3' base=None pins=['DK', 'G', 'S', 'B', 'DA', 'A', 'DB', 'D1', 'G1', 'D2', 'G2'] parameters={} prefix='X' elements=List[Instance](__root__=[Instance(model='PLVT', name='MN1', pins={'D': 'A', 'G': 'B', 'S': 'C', 'B': 'GND'}, parameters={'W': '2.7E-07', 'L': '1', 'NFIN': '1', 'NF': '4', 'M': '3', 'STACK': '1', 'PARALLEL': '1'}, generator='PMOS', abstract_name=None), Instance(model='PSVT', name='MI1', pins={'D': 'DA', 'G': 'DA', 'S': 'S', 'B': 'B'}, parameters={'W': '1.8E-07', 'L': '1', 'NFIN': '1', 'NF': '1', 'M': '1', 'STACK': '3', 'PARALLEL': '1'}, generator='PMOS', abstract_name=None), Instance(model='DP_NMOS_B', name='X_DP_NMOS_B_MN2_MN3', pins={'DA': 'D1', 'GA': 'G1', 'S': 'S', 'B': 'GND', 'DB': 'D2', 'GB': 'G2'}, parameters=None, generator='DP_NMOS_B', abstract_name=None), Instance(model='DP_PMOS_B', name='X_DP_PMOS_B_MI0_MI2', pins={'DA': 'DK', 'GA': 'G', 'S': 'S', 'B': 'B', 'DB': 'DA', 'GB': 'DB'}, parameters=None, generator='DP_PMOS_B', abstract_name=None)]) constraints=ConstraintDB(__root__=[])
FAILED

============================================================================================ FAILURES =============================================================================================
__________________________________________________________________________________ test_sizing3[intel_circuit3] ___________________________________________________________________________________

ckt = SubCircuit(name='INTEL_CIRCUIT3', base=None, pins=['DK', 'G', 'S', 'B', 'DA', 'A', 'DB', 'D1', 'G1', 'D2', 'G2'], para...DA', 'GB': 'DB'}, parameters=None, generator='DP_PMOS_B', abstract_name=None)]), constraints=ConstraintDB(__root__=[]))

    @pytest.mark.parametrize('cn',['intel_circuit3'])
    def test_sizing3(ckt):
        assert ckt.elements[3]
>       assert len(ckt.elements) == 5, f"{[ele.name for ele in ckt.elements]}"
E       AssertionError: ['MN1', 'MI1', 'X_DP_NMOS_B_MN2_MN3', 'X_DP_PMOS_B_MI0_MI2']
E       assert 4 == 5
E         +4
E         -5

tests/compiler/test_sizing.py:53: AssertionError
--------------------------------------------------------------------------------------- Captured log setup ----------------------------------------------------------------------------------------
WARNING  align.compiler.preprocess:preprocess.py:160 No power and gnd in this circuit INTEL_CIRCUIT3, please check setup file
----------------------------------------------------------------- generated xml file: /home/smburns/DARPA/ALIGN-public/junit.xml ------------------------------------------------------------------
===================================================================================== short test summary info =====================================================================================
FAILED tests/compiler/test_sizing.py::test_sizing3[intel_circuit3] - AssertionError: ['MN1', 'MI1', 'X_DP_NMOS_B_MN2_MN3', 'X_DP_PMOS_B_MI0_MI2']
================================================================================= 1 failed, 4 deselected in 0.29s =================================================================================
```
We can try to exclude this test if z3-solver can't be found, or we can make it work even if it is not available.
For some old environments at Intel we can't install z3-solver, but if we are being to rely on z3-solver, then we can remove this old environment from CI.
Let me know if you can replicate this:
```
pip uninstall z3-solver
```

kkunal1408 commented:
Thanks Steve,
I will modify it to avoid the z3 checker.

Kunal

On Thu, Aug 19, 2021 at 3:27 PM Steve Burns ***@***.***>
wrote:

> It seems that the z3-solver isn't installed on that particular setup and
> your code depends on it being installed to produce the correct answer.
>
> This is what I get on my local machine when I uninstall z3-solver:
>
> (.venv) ***@***.***:~/DARPA/ALIGN-public$ pytest -s -k test_sizing3 -vv -- tests/compiler/test_sizing.py
> align.schema.checker WARNING : Could not import z3. Z3Checker disabled.
> ======================================================================================= test session starts =======================================================================================
> platform linux -- Python 3.8.0, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/smburns/DARPA/ALIGN-public/.venv/bin/python3.8
> cachedir: .pytest_cache
> rootdir: /home/smburns/DARPA/ALIGN-public, configfile: pytest.ini
> plugins: cov-2.12.1, forked-1.3.0, timeout-1.4.2, cpp-2.0.0, dash-1.20.0, xdist-2.2.1
> collected 5 items / 4 deselected / 1 selected
>
> tests/compiler/test_sizing.py::test_sizing3[intel_circuit3] align.compiler.preprocess WARNING : No power and gnd in this circuit INTEL_CIRCUIT3, please check setup file
> intel_circuit3
> name='INTEL_CIRCUIT3' base=None pins=['DK', 'G', 'S', 'B', 'DA', 'A', 'DB', 'D1', 'G1', 'D2', 'G2'] parameters={} prefix='X' elements=List[Instance](__root__=[Instance(model='PLVT', name='MN1', pins={'D': 'A', 'G': 'B', 'S': 'C', 'B': 'GND'}, parameters={'W': '2.7E-07', 'L': '1', 'NFIN': '1', 'NF': '4', 'M': '3', 'STACK': '1', 'PARALLEL': '1'}, generator='PMOS', abstract_name=None), Instance(model='PSVT', name='MI1', pins={'D': 'DA', 'G': 'DA', 'S': 'S', 'B': 'B'}, parameters={'W': '1.8E-07', 'L': '1', 'NFIN': '1', 'NF': '1', 'M': '1', 'STACK': '3', 'PARALLEL': '1'}, generator='PMOS', abstract_name=None), Instance(model='DP_NMOS_B', name='X_DP_NMOS_B_MN2_MN3', pins={'DA': 'D1', 'GA': 'G1', 'S': 'S', 'B': 'GND', 'DB': 'D2', 'GB': 'G2'}, parameters=None, generator='DP_NMOS_B', abstract_name=None), Instance(model='DP_PMOS_B', name='X_DP_PMOS_B_MI0_MI2', pins={'DA': 'DK', 'GA': 'G', 'S': 'S', 'B': 'B', 'DB': 'DA', 'GB': 'DB'}, parameters=None, generator='DP_PMOS_B', abstract_name=None)]) constraints=ConstraintDB(__root__=[])
> FAILED
>
> ============================================================================================ FAILURES =============================================================================================
> __________________________________________________________________________________ test_sizing3[intel_circuit3] ___________________________________________________________________________________
>
> ckt = SubCircuit(name='INTEL_CIRCUIT3', base=None, pins=['DK', 'G', 'S', 'B', 'DA', 'A', 'DB', 'D1', 'G1', 'D2', 'G2'], para...DA', 'GB': 'DB'}, parameters=None, generator='DP_PMOS_B', abstract_name=None)]), constraints=ConstraintDB(__root__=[]))
>
>     @pytest.mark.parametrize('cn',['intel_circuit3'])
>     def test_sizing3(ckt):
>         assert ckt.elements[3]
> >       assert len(ckt.elements) == 5, f"{[ele.name for ele in ckt.elements]}"
> E       AssertionError: ['MN1', 'MI1', 'X_DP_NMOS_B_MN2_MN3', 'X_DP_PMOS_B_MI0_MI2']
> E       assert 4 == 5
> E         +4
> E         -5
>
> tests/compiler/test_sizing.py:53: AssertionError
> --------------------------------------------------------------------------------------- Captured log setup ----------------------------------------------------------------------------------------
> WARNING  align.compiler.preprocess:preprocess.py:160 No power and gnd in this circuit INTEL_CIRCUIT3, please check setup file
> ----------------------------------------------------------------- generated xml file: /home/smburns/DARPA/ALIGN-public/junit.xml ------------------------------------------------------------------
> ===================================================================================== short test summary info =====================================================================================
> FAILED tests/compiler/test_sizing.py::test_sizing3[intel_circuit3] - AssertionError: ['MN1', 'MI1', 'X_DP_NMOS_B_MN2_MN3', 'X_DP_PMOS_B_MI0_MI2']
> ================================================================================= 1 failed, 4 deselected in 0.29s =================================================================================
>
> We can try to exclude this test if z3-solver can't be found, or we can
> make it work even if it is not available.
> For some old environments at Intel we can't install z3-solver, but if we
> are being to rely on z3-solver, then we can remove this old environment
> from CI.
> Let me know if you can replicate this:
>
> pip uninstall z3-solver
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ALIGN-analoglayout/ALIGN-public/pull/615#issuecomment-902293329>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJUB53JUBYGVETKQRNEU54DT5WAN3ANCNFSM42W64INQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>
-- 
Sent from mobile. Excuse for any typos.

stevenmburns commented:
@kkunal1408 I'm leaning toward removing support for this old platform. If we are using z3-solver now to do non-trivial things than we need it, and should remove platforms where we can't make it work. The old platform at Intel is not really needed. It was a nice to have that we can skip if that is the right thing to do.

stevenmburns reviewed (approved):
I think we can check this in if you think it is okay.

stevenmburns commented:
@parijatm @soneryaldiz I removed the CI run for manylinux1 since we couldn't get z3-solver to work on that platform. It is my understand that we only wanted to support the platform so we could run ALIGN on SLES11 machines at Intel. I think the need for that has gone away. Let me know if I'm wrong.

I think that if we want to use the z3-solver in a significant way in ALIGN that is more important than supporting SLES11.

stevenmburns commented:
@kkunal1408 Several of the nightly blocks are still failing. We might want to look at these in another PR
```
===================================================================================== short test summary info ======================================================================================
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-vco_dtype_12_hierarchical] - AssertionError: vco_dtype_12_hierarchical :No results generated
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-powertrain_binary]
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-test_vga]
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-COMPARATOR_2LEVEL_BIDIRECTIONAL_MAC_SKEW] - AssertionError: COMPARATOR_2LEVEL_BIDIRECTIONAL_MAC_SKEW :No result...
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-Sanitized_Coarse_Comp_CK] - AssertionError: Sanitized_Coarse_Comp_CK :No results generated
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-COMP_GM_STAGE_0415] - AssertionError: COMP_GM_STAGE_0415 :No results generated
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-telescopic_ota_guard_ring] - AssertionError: telescopic_ota_guard_ring :No results generated
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-vco_dtype_12_hierarchical_res] - AssertionError: vco_dtype_12_hierarchical_res :No results generated
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-CTDSM_CORE_NEW] - AssertionError: CTDSM_CORE_NEW :No results generated
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-vco_dtype_12_hierarchical_res_constrained] - AssertionError: vco_dtype_12_hierarchical_res_constrained :No resu...
FAILED tests/integration/test_all.py::test_A[20-top_down-True-FinFET14nm_Mock_PDK-SW_Cres_v3_5] - AssertionError: SW_Cres_v3_5 :No results generated
```

parijatm commented:
> It is my understand that we only wanted to support the platform so we could run ALIGN on SLES11 machines at Intel.

That's correct. Let's remove it. We can always build the wheel outside of CI if we really really need it.

There was a second reason for building the manylinux1 wheel with every commit. It's actually the python version (3.7) I think we wanted to keep python3.7 compatibility as it's the default on Ubuntu 18.04.

If that's the case I would change the manylinux1 test to a manylinux2010 test instead of removing it altogether. Many of us use python3.8 and may inadvertently end up using python3.8+ features.

stevenmburns commented:
> > It is my understand that we only wanted to support the platform so we could run ALIGN on SLES11 machines at Intel.
> 
> That's correct. Let's remove it. We can always build the wheel outside of CI if we really really need it.
> 
> There was a second reason for building the manylinux1 wheel with every commit. It's actually the python version (3.7) I think we wanted to keep python3.7 compatibility as it's the default on Ubuntu 18.04.
> 
> If that's the case I would change the manylinux1 test to a manylinux2010 test instead of removing it altogether. Many of us use python3.8 and may inadvertently end up using python3.8+ features.

Thanks @parijatm , I can add that combination to CI.

stevenmburns commented:
@kkunal1408 

The three examples `powertrain`, `powertrain_binary` and `powertrain_thermo` are all failing because they have no GND port.
Here is the stack trace for `powertrain`:
```
align.cmdline ERROR : Fatal Error. Cannot proceed
Traceback (most recent call last):
  File "/home/smburns/DARPA/ALIGN-public/align/cmdline.py", line 142, in parse_args
    return schematic2layout(**vars(arguments))
  File "/home/smburns/DARPA/ALIGN-public/align/main.py", line 329, in schematic2layout
    primitives = generate_hierarchy(netlist, subckt, topology_dir, flatten, pdk_dir, uniform_height)
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 25, in generate_hierarchy
    ckt_data = compiler_input(netlist_path, subckt, pdk_dir, config_path, flatten_heirarchy)
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 109, in compiler_input
    preprocess_stack_parallel(ckt_data, design_setup, design_name.upper())
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/preprocess.py", line 39, in preprocess_stack_parallel
    define_SD(top, design_setup['POWER'], design_setup['GND'], design_setup['DIGITAL'])
  File "/home/smburns/DARPA/ALIGN-public/align/compiler/preprocess.py", line 172, in define_SD
    assert low, f"GND port {gnd} not defined in subckt port {circuit.pins}"
AssertionError: GND port ['VSS'] not defined in subckt port ['VCC', 'VG', 'VOUT']
```

kkunal1408 commented:
> @kkunal1408
> 
> The three examples `powertrain`, `powertrain_binary` and `powertrain_thermo` are all failing because they have no GND port.
> Here is the stack trace for `powertrain`:
> 
> ```
> align.cmdline ERROR : Fatal Error. Cannot proceed
> Traceback (most recent call last):
>   File "/home/smburns/DARPA/ALIGN-public/align/cmdline.py", line 142, in parse_args
>     return schematic2layout(**vars(arguments))
>   File "/home/smburns/DARPA/ALIGN-public/align/main.py", line 329, in schematic2layout
>     primitives = generate_hierarchy(netlist, subckt, topology_dir, flatten, pdk_dir, uniform_height)
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 25, in generate_hierarchy
>     ckt_data = compiler_input(netlist_path, subckt, pdk_dir, config_path, flatten_heirarchy)
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/compiler.py", line 109, in compiler_input
>     preprocess_stack_parallel(ckt_data, design_setup, design_name.upper())
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/preprocess.py", line 39, in preprocess_stack_parallel
>     define_SD(top, design_setup['POWER'], design_setup['GND'], design_setup['DIGITAL'])
>   File "/home/smburns/DARPA/ALIGN-public/align/compiler/preprocess.py", line 172, in define_SD
>     assert low, f"GND port {gnd} not defined in subckt port {circuit.pins}"
> AssertionError: GND port ['VSS'] not defined in subckt port ['VCC', 'VG', 'VOUT']
> ```
One solution for this error is to remove the port from the setup file. Though not sure of the intent behind the port in the setup file but not in design. Do we want to create a power grid of a net in case the net is not there in the design?

kkunal1408 commented:
@parijatm in the vco_dtype_hierarchical the parser does not understand a next line operator "\".
e.g. .param nfin=14 rres=2k lastt=60n fin_n_diff2sing=4 fin_p_diff2sing=6 \
    width_n_diff2sing=10
Is there simple way to fix it in the regex?

parijatm commented:
> @parijatm in the vco_dtype_hierarchical the parser does not understand a next line operator "\".
> e.g. .param nfin=14 rres=2k lastt=60n fin_n_diff2sing=4 fin_p_diff2sing=6 \
> width_n_diff2sing=10
> Is there simple way to fix it in the regex?

I thought next line operator in spice was "+" at the beginning of the line. I think I did implement that. If so, you may look at that code to see how "\" can be implemented (first do check to see its valid spice). I dont have access to my laptop at the moment.

stevenmburns commented:
@kkunal1408 Fixed up two of the powertrain cases. The third (powertrain_binary) is still failing.

kkunal1408 commented:
Hi Steve, 
I will fix the powertrain_binary, I needed help with parsing the "vco_dtype_12_hierarchical" . The netlist has "\\" at end of line which means a continue statement but I could not figure out the right regex syntax to add this type of parsing (schema/parser.py).

stevenmburns commented:
@kkunal1408 That should allow continuation characters. The vco_dtype cases still don't work because (at least) the parameter definitions in a subckt (without an '=' value) are causing errors.
